lib/linter/linter.js

```diff
@@ -866,93 +868,42 @@ function analyzeScope(ast, languageOptions, visitorKeys) {
-            eslintVisitorKeys: true,
-            eslintScopeManager: true,
-            filePath
-        }
-    );
-
-    /*
-     * Check for parsing errors first. If there's a parsing error, nothing
-     * else can happen. However, a parsing error does not throw an error
-     * from this method - it's just considered a fatal error message, a
-     * problem that ESLint identified just like any other.
-     */
-    try {
-        debug("Parsing:", filePath);
-        const parseResult = (typeof parser.parseForESLint === "function")
-            ? parser.parseForESLint(textToParse, parserOptions)
-            : { ast: parser.parse(textToParse, parserOptions) };
-
-        debug("Parsing successful:", filePath);
-        const ast = parseResult.ast;
-        const parserServices = parseResult.services || {};
-        const visitorKeys = parseResult.visitorKeys || evk.KEYS;
-
-        debug("Scope analysis:", filePath);
-        const scopeManager = parseResult.scopeManager || analyzeScope(ast, languageOptions, visitorKeys);
+function parse(file, language, languageOptions) {
 
-        debug("Scope analysis successful:", filePath);
+    const result = language.parse(file, { languageOptions });
 
+    if (result.ok) {
         return {
             success: true,
-
-            /*
-             * Save all values that `parseForESLint()` returned.
-             * If a `SourceCode` object is given as the first parameter instead of source code text,
-             * linter skips the parsing process and reuses the source code object.
-             * In that case, linter needs all the values that `parseForESLint()` returned.
-             */
-            sourceCode: new SourceCode({
-                text,
-                ast,
-                parserServices,
-                scopeManager,
-                visitorKeys
-            })
-        };
-    } catch (ex) {
-
-        // If the message includes a leading line number, strip it:
-        const message = `Parsing error: ${ex.message.replace(/^line \d+:/iu, "").trim()}`;
-
-        debug("%s\n%s", message, ex.stack);
-
-        return {
-            success: false,
-            error: {
-                ruleId: null,
-                fatal: true,
-                severity: 2,
-                message,
-                line: ex.lineNumber,
-                column: ex.column,
-                nodeType: null
-            }
+            sourceCode: language.createSourceCode(file, result, { languageOptions })
         };
     }
+
+    // if we made it to here there was an error
+    return {
+        success: false,
+        errors: result.errors.map(error => ({
+            ruleId: null,
+            nodeType: null,
+            fatal: true,
+            severity: 2,
+            message: error.message,
+            line: error.line,
+            column: error.column,
+            endLine: error.endLine || error.line,
+            endColumn: error.endColumn || error.column + 1
```

Comment:
Parsing error can be at the end of a line, or even at the end of the file (e.g., when the whole file is `(`). In these cases, `error.column + 1` would be an invalid location. I think it's fine to omit `endLine` and `endColumn` if the parser doesn't provide them (i.e., keep the current behavior).

Reply:
I much prefer having an `endLine` and `endColumn` always present for consistency. I don't think it's a problem to have a number past the end of the text (by 1).

Reply:
A column that is last+1 is a valid location that represents end of line or end of file. But when parser returns that column as the start and we increment it by one to calculate the end, the end column will be last+2, which is invalid.

In the `(` example, Espree returns { line: 1, column: 1 }, which we translate to ( line: 1, column: 2 } to return a location with a 1-based column index. That's a valid location right after the `(`. But if we return ( line: 1, column: 2, endLine: 1, endColumn: 3 }, the end location will be invalid.

Some IDEs/tools handle invalid locations well (e.g., VS Code), but some IDEs/tools may be more strict. I recall that Atom wasn't even showing the lint error message, but just an "Invalid position" text.

https://github.com/AtomLinter/linter-eslint/issues?page=1&q=is%3Aissue+is%3Aclosed+%22invalid+position%22+in%3Atitle

Reply:
Okay, I'll revert this.